pennylane/math/single_dispatch.py

```diff
@@ -0,0 +1,327 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Autoray registrations"""
+# pylint:disable=protected-access,import-outside-toplevel,wrong-import-position
+from importlib import import_module
+import numbers
+
+import autoray as ar
+import numpy as np
+
+
+def _i(name):
+    """Convenience function to import PennyLane
+    interfaces via a string pattern"""
+    if name == "tf":
+        return import_module("tensorflow")
+
+    if name == "qml":
+        return import_module("pennylane")
+
+    return import_module(name)
+
+
+# -------------------------------- NumPy --------------------------------- #
+from scipy.linalg import block_diag as _scipy_block_diag
+
+ar.register_function("numpy", "flatten", lambda x: x.flatten())
+ar.register_function("numpy", "coerce", lambda x: x)
+ar.register_function("numpy", "block_diag", lambda x: _scipy_block_diag(*x))
+ar.register_function("builtins", "block_diag", lambda x: _scipy_block_diag(*x))

```

Comment:
For my understanding, how come we are registering on `"builtins"`? How would the syntax look afterwards, would we have a `block_diag` name available in the scope?

Reply:
Oh, great question. `builtins` simply means that the first argument to the function is a Python built-in type, such as a list or a tuple. For example:

```python
autoray.numpy.sum([0, 2, 3])  # will dispatch to the 'builtins' registered function
autoray.numpy.sum(np.array([0, 2, 3]))  # will dispatch to the 'numpy' registered function
```

This is quite nice, as it provides us very fine control - we could even choose to have a `"builtins"` function use `cmath` for speed as opposed to NumPy.